conu/apidefs/container.py

```diff
@@ -0,0 +1,234 @@
+"""
+Abstract class definitions for containers.
+"""
+
+# FIXME: TBD
+from .image import Image
+
+
+class BinaryArgsBuilder(object):
+    """
+    class which helps you build a command so it can be executed via subprocess
+
+    TBD: I think I
+    """
+    template = "{binary} {global_opts} {cmd} {opts} {args}"
+    binary = ""
+    global_options = ""
+    command = ""
+    arguments = ""
+    options = ""
+
+    def build(self):
+        command = self.template.format(
+            binary=self.binary,
+            global_opts=self.global_options,
+            cmd=self.command,
+            opts=self.options,
+            args=self.arguments
+        )
+        return command
+
+
+class ContainerParameters(object):
+    """
+    key value store of container parameters
+
+    TBD
+    """
+
+
+class Container(object):
+    """
+    Container class definition which contains abstract methods. The instances should call the
+    constructor
+    """
+    def __init__(self, image, container_id):
+        """
+        :param image: Image instance
+        :param container_id: str, unique identifier of this container
+        """
+        if not isinstance(image, Image):
+            raise RuntimeError("image argument is not an instance of Image class")
+        self.image = image
+        self.container_id = container_id
+        self._metadata = None
+
+    def get_metadata(self, refresh=False):
+        """
+        return cached metadata by default
+
+        :param refresh: bool, returns up to date metadata if set to True
+        :return: dict
+        """
+        raise NotImplementedError("get_metadata method is not implemented")
+
+    def is_running(self):
+        """
+        returns True if the container is running, this method should always ask the API and
+        should not use a cached value
+
+        :return: bool
+        """
+        raise NotImplementedError("is_running method is not implemented")
+
+    def status(self):
+        """
+        Provide current, up-to-date status of this container. This method should not use cached
+        value. Implementation of this method should clearly state list of possible values
+        to get from this method
+
+        :return: str
+        """
+        raise NotImplementedError("status method is not implemented")
+
+    def get_pid(self):
+        """
+        get process identifier of the root process in the container
+
+        :return: int
+        """
+        raise NotImplementedError("get_pid method is not implemented")
+
+    def name(self):
+        """
+        Return name of this container.
+
+        :return: str
+        """
+        raise NotImplementedError("name method is not implemented")
+
+    def get_ip(self):
```

Comment:
I also think that we should have `get_ip4` and `get_ip6`; `get_ip` should then return the same as `get_ip4`.

Reply:
I agree this is a good design: I fear it won't be that easy to implement it.

Reply:
So maybe we just want to have a `get_IP4` or similar, that would return list of IPv4 addresses and we don't need to care about the rest ATM. Let's just design the API in a way that will allow us to do this in future if we need to.